source/common/common/assert.h

```diff
@@ -134,6 +160,32 @@ void invokeDebugAssertionFailureRecordAction_ForAssertMacroUseOnly();
     abort();                                                                                       \
   } while (false)
 
+// CONDITION_STR is needed to prevent macros in condition from being expected, which obfuscates
+// the logged failure, e.g., "EAGAIN" vs "11".
+#define _ENVOY_BUG_IMPL(CONDITION, CONDITION_STR, ACTION, DETAILS)                                 \
+  do {                                                                                             \
+    if (!(CONDITION)) {                                                                            \
+      const std::string& details = (DETAILS);                                                      \
+      ENVOY_LOG_TO_LOGGER(Envoy::Logger::Registry::getLog(Envoy::Logger::Id::envoy_bug), error,    \
+                          "envoy bug failure: {}.{}{}", CONDITION_STR,                             \
+                          details.empty() ? "" : " Details: ", details);                           \
```

Comment:
drive by: I think you want the logging to also be in action / back off code. Logging at high rate will topple the server. Personally, I would probably log with exponential backoff, but stat on every instance, since stats are super cheap.

Reply:
+1 we want exponential backoff on generated logs.

Reply:
done -- implemented inside the macro just to avoid propogating DETAILS around.